### PR TITLE
68 69 sandbox candidate api

### PR DESCRIPTION
**Pull request by MikitaYepishkin**

### CODE REVIEW

File: SAPex/Controllers/SandboxesController.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                return await Task.FromResult(NotFound());
+            }
+
+            SandboxViewModel viewModel = _mapper.MapSbStackLgFromDtoToView(sandboxDtoModel, languageDtoModel, stackTechnologyDtoModel);
+
+            return await Task.FromResult(Ok(viewModel)); // convert to json?
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> GetByPage(int pageNumber = 1, int pageSize = 5, string sortOrder = "name", string searchString = "")
```
Comment (dnavitski) [inline SAPex/Controllers/SandboxesController.cs]: Please use one ViewModel with fields as an input parameter.

File: SAPex/Controllers/SandboxesController.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            IEnumerable<SandboxDtoModel> dtoModels = await _sandboxService.GetAllSandboxesAsync(); // on what level goes null check? it should go directly to db?
+
+            // IEnumerable<IEnumerable<StackTechnologyDtoModel>> stackTechnologyDtoModels =
+            // IEnumerable<LanguageDtoModel> languageDtoModel = await _languageService.GetLanguagesBySandboxIdAsync(id);
+
+            if (dtoModels == null)
+            {
+                return await Task.FromResult(NotFound()); // return null
```
Comment (dnavitski) [inline SAPex/Controllers/SandboxesController.cs]: Please don't use NotFound ( ) for search keys.

File: SAPex/Controllers/SandboxesController.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            }
+
+            IEnumerable<SandboxViewModel> viewModels = _mapper.MapListSbFromDtoToView(dtoModels);
+
+            if (!string.IsNullOrEmpty(searchString))
```
Comment (dnavitski) [inline SAPex/Controllers/SandboxesController.cs]: Please remove this functionality into the Repository layer or comment.

File: SAPex/Controllers/StackTechnologiesController.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> GetAll()
+        {
+            IEnumerable<StackTechnologyDtoModel> dtoModels = await _service.GetAllStackTechnologiesAsync(); // on what level goes null check? it should go directly to db?
+
+            if (dtoModels == null)
+            {
+                return await Task.FromResult(NotFound()); // return null
```
Comment (dnavitski) [inline SAPex/Controllers/StackTechnologiesController.cs:49]: Please don't use NotFound() method for search or GetAll cases.

**Review by dnavitski — CHANGES_REQUESTED** [review]

File: SAPex/SAPex.csproj [hunk 5]
```diff
@@ -11,20 +11,28 @@
...
   <ItemGroup>
+    <PackageReference Include="FluentValidation.AspNetCore" Version="10.3.4" />
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="5.0.11" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="5.0.11" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="5.0.11">
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
     <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="5.0.2" />
+    <PackageReference Include="PagedList.Mvc" Version="4.5.0" />
```
Comment (dnavitski) [inline SAPex/SAPex.csproj]: Please exclude this library from the project.

**Review by dnavitski — CHANGES_REQUESTED** [review]

File: SAPex/Controllers/SandboxesController.cs (hunk 1 above)
Comment (MaximPyatigor) [reply SAPex/Controllers/SandboxesController.cs]: have little problem with passing parameter to method

File: SAPex/Controllers/SandboxesController.cs (hunk 2 above)
Comment (MaximPyatigor) [reply SAPex/Controllers/SandboxesController.cs]: removed

File: SAPex/Controllers/SandboxesController.cs (hunk 3 above)
Comment (MaximPyatigor) [reply SAPex/Controllers/SandboxesController.cs]: done

File: SAPex/Controllers/StackTechnologiesController.cs (hunk 4 above)
Comment (MaximPyatigor) [reply SAPex/Controllers/StackTechnologiesController.cs:49]: removed

File: SAPex/SAPex.csproj (hunk 5 above)
Comment (MaximPyatigor) [reply SAPex/SAPex.csproj]: done

File: SAPex/Controllers/LanguagesControllerTest.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿using Microsoft.AspNetCore.Mvc;
+using SAPex.Controllers;
+using SAPex.Models;
+
+namespace SAPex
+{
+    [Route("api/languagesTest")]
+    [ApiController]
+    public class LanguagesControllerTest : AbstractNameController<LanguageViewModelTest>
```
Comment (dnavitski) [inline SAPex/Controllers/LanguagesControllerTest.cs:9]: @MikitaYepishkin , @MaximPyatigor 
Why and where is this controller used?

File: SAPex/Controllers/LanguagesController.cs [hunk 7]
```diff
@@ -1,12 +1,56 @@
...
+
+            return await Task.FromResult(Ok(viewModel));
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> GetAll()
+        {
+            IEnumerable<LanguageDtoModel> dtoModels = await _service.GetAllLanguagesAsync();
+
+            if (dtoModels == null)
```
Comment (dnavitski) [inline SAPex/Controllers/LanguagesController.cs:46]: Please remove this check case.

File: SAPex/Controllers/SandboxesController.cs [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            _languageService = languageService;
+
+            _inputParamersMapper = new InputParametrsMapper();
+        }
+
+        [HttpGet("{id}")]
+        public async Task<IActionResult> Get([FromRoute] Guid id)
+        {
+            SandboxDtoModel sandboxDtoModel = await _sandboxService.FindSandboxByIdAsync(id); // on what level goes null check? it should go directly to db?
+            IEnumerable<StackTechnologyDtoModel> stackTechnologyDtoModel = await _stackTechnologyService.GetStackTechnologiesBySandboxIdAsync(id);
```
Comment (dnavitski) [inline SAPex/Controllers/SandboxesController.cs:49]: If we use a variable as a single item, this variable will be named 'stackTechnology';
If we use a variable as a list of elements, this variable should be named like 'stackTechnologies'.

File: BusinessLogicLayer/Services/SandboxService.cs [hunk 9]
```diff
@@ -58,6 +54,19 @@ public async Task<IEnumerable<SandboxDtoModel>> GetAllSandboxesAsync()
             return profile.mapListToDto(sandboxList);
         }
 
+        public async Task<PagedList<SandboxDtoModel>> GetPagedSandboxesAsync(InputParametrsDtoModel parametrs)
+        {
+            PagedList<SandboxEntityModel> sandboxList = await Task.Run(() => unitOfWork.Sandboxes.GetPagedAsync(inputParametrsProfile.MapFromDtoToEntity(parametrs)));
```
Comment (dnavitski) [inline BusinessLogicLayer/Services/SandboxService.cs:59]: Please don't use Task.Run() for async methods.

**Review by dnavitski — APPROVED** [review]

**Review by dnovitsky — APPROVED** [review]

File: SAPex/Controllers/SandboxesController.cs (hunk 1 above)
Comment (MaximPyatigor) [reply SAPex/Controllers/SandboxesController.cs]: done

**Pull request merged**